simplified-app-shared/src/main/java/org/nypl/simplified/app/AnnotationsManager.kt

```diff
@@ -0,0 +1,600 @@
+import org.nypl.simplified.app.catalog.annotation.AnnotationResult
+import org.nypl.simplified.app.reader.ReaderBookLocation
+import org.nypl.simplified.books.core.AccountCredentials
+import org.nypl.simplified.books.core.AccountsControllerType
+import org.nypl.simplified.multilibrary.Account
+import org.nypl.simplified.volley.NYPLJsonObjectRequest
+import org.slf4j.LoggerFactory
+import java.io.IOException
+
+/**
+ * Performs work relevant to syncing OPDS Annotations.
+ * Current features for books:
+ * 1 - Current reading position
+ * 2 - Bookmarks saved by the user
+ */
+class AnnotationsManager(private val libraryAccount: Account,
+                         private val credentials: AccountCredentials,
+                         private val context: Context)
+{
+  private val requestQueue = Volley.newRequestQueue(context)
+
+  private companion object {
+    val LOG = LoggerFactory.getLogger(AnnotationsManager::class.java)!!
+  }
+
+  /**
+   * Queries the server, and based on a set of conditions, decides whether
+   * or not to inform the user about "Syncing" in an Alert Dialog, where
+   * the user can choose to dismiss/ignore, or turn it on for the first time.
+   * @param account The particular library for the permission status request.
+   * @param completion Asynchronous handler for the user's decision, or error.
+   */
+  fun requestServerSyncPermissionStatus(account: AccountsControllerType,
+                                        completion: (enableSync: Boolean) -> Unit)
+  {
+    if (!syncIsPossible(account)) {
+      LOG.debug("Account does not satisfy conditions for sync setting request.")
+      completion(false)
+      return
+    }
+
+    //TODO I think this condition is either wrong or out of date.. think about it some more during testing
+    if (Simplified.getSharedPrefs().getBoolean("userHasSeenFirstTimeSyncMessage") == true &&
+        Simplified.getSharedPrefs().getBoolean("syncPermissionGranted", libraryAccount.id) == false) {
+      completion(false)
+      return
+    }
+
+    syncPermissionStatusUriRequest { initialized, syncIsPermitted ->
+      if (initialized && syncIsPermitted) {
+        Simplified.getSharedPrefs().putBoolean("userHasSeenFirstTimeSyncMessage", true)
+        LOG.debug("Sync has already been enabled on the server. Enable here as well.")
+        completion(true)
+      } else if (!initialized && !Simplified.getSharedPrefs().getBoolean("userHasSeenFirstTimeSyncMessage")) {
+        LOG.debug("Sync has never been initialized for the patron. Proceeding with opt-in alert.")
+        presentFirstTimeSyncAlertDialog(completion)
+      } else {
+        LOG.debug("Sync has been initialized and not permitted. Continuing with sync disabled.")
+        completion(false)
+      }
+    }
+  }
+
+  private fun syncPermissionStatusUriRequest(completion: (initialized: Boolean, syncIsPermitted: Boolean) -> Unit)
+  {
+    val catalogUriString = libraryAccount.catalogUrl ?: null
```

Comment:
do we need this `?:`? I thought the Elvis operator would only return the right side of the expression if the left is `null` already

Reply:
This is a good question and Winnie and I were discussing this as well. The compiler gives me the inferred type which is this "platform type" of "either nullable or nonnull"... which makes me kind of nervous. My first instinct as to simply failably cast "as?" but the compiler says "don't need that"... 

I will be reviewing these areas that I use the elvis for this purpose and I think the Idea that JetBrains is pushing to to just look downstream and protect for NPE's where appropriate.. I just don't have an intuitive feel for that yet.

Reply:
cool

Reply:
I might do `val baseUri = libraryAccount.catalogUrl?.let { Uri.parse(it)}`, but just personal preference

Reply:
Thanks David. I read up on extension functions a little bit. I'll keep an eye out for more places to use this.